File: 1_Algorithmic_Toolbox/week5_dynamic_programming1/1_money_change_again/ChangeDP.java
import java.util.Scanner;

public class ChangeDP {
    private static final int COINS[] = { 1, 3, 4 };

    private static int getChange(int m) {
        // write your code here
        int minCoinCounts[] = new int[m + 1];
        minCoinCounts[0] = 0;
        for (int i = 1; i < (m + 1); ++i) {
            minCoinCounts[i] = Integer.MAX_VALUE;
            for (int coin : COINS) {
                if (i >= coin) {
                    int coinCount = minCoinCounts[i - coin] + 1;
                    minCoinCounts[i] = Math.min(coinCount, minCoinCounts[i]);
                }
            }
        }
        return minCoinCounts[m];
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int m = scanner.nextInt();
        System.out.println(getChange(m));
    }
}

File: 1_Algorithmic_Toolbox/week5_dynamic_programming1/1_money_change_again/change_dp.py
# Uses python3
import sys
DEBUG = False
# def get_change(m):
#     #write your code here
#     return m // 4

# write a dynamic programming solution for get_change on 
# the coins [1, 3, 4]
def get_change(m):
    """
    Calculates the minimum number of coins needed to make change for a given amount.

    Parameters:
    m (int): The amount for which change needs to be made.

    Returns:
    int: The minimum number of coins needed to make change for the given amount.

    Examples:
    >>> get_change(2)
    2
    >>> get_change(34)
    9
    
    """
    coins = [1, 3, 4]
    # initialize min_num_coins
    min_num_coins = [0] + [float('inf')] * m
    for i in range(1, m + 1):
        for coin in coins:
            if i >= coin:
                num_coins = min_num_coins[i - coin] + 1
                if num_coins < min_num_coins[i]:
                    min_num_coins[i] = num_coins
    return min_num_coins[m]

# write test_case
def test_case():
    """
    Some test cases for get_change function.

    Parameters:
    """
    assert get_change(2) == 2
    assert get_change(34) == 9
    print("Test passed...")
    return


if __name__ == '__main__':
    if DEBUG:
        test_case()
    m = int(sys.stdin.read())
    print(get_change(m))
